File: server/NaverNews.py
import requests, pandas as pd, certifi, bs4
from bs4 import BeautifulSoup
from io import StringIO
from urllib.parse import urlparse, parse_qs, urljoin

USER_AGENT = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)"}

def get_news_tables(code: str, page: int = 1):
    base = f"https://finance.naver.com/item/news.naver?code={code}"
    sess = requests.Session()
    sess.headers.update({
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
        "Accept-Language": "ko-KR,ko;q=0.9,en-US;q=0.8",
    })

    # 1️⃣ 첫 페이지 요청 → 쿠키 획득
    r1 = sess.get(base, verify=certifi.where(), timeout=10)
    r1.raise_for_status()

    # 2️⃣ iframe src 추출
    soup = BeautifulSoup(r1.text, "html.parser")
    src = soup.select_one("#news_frame")["src"]
    if "?page=" not in src:          # 혹시 page 파라미터가 비어 있으면 채우기
        src += f"&page={page}"
    iframe_url = urljoin(base, src)

    # 3️⃣ 같은 세션 + Referer 로 두 번째 요청
    r2 = sess.get(
        iframe_url,
        headers={"Referer": base, **sess.headers},
        verify=certifi.where(),
        timeout=10,
    )
    r2.raise_for_status()

    soup = bs4.BeautifulSoup(r2.text, "html.parser")
    for tr in soup.select("tr.relation_lst"):
        tr.decompose()                       # DOM에서 완전히 삭제

    html_clean = str(soup)

    # print(html_clean)

    links = []
    for a in soup.find_all("a", href=True):
        href = a["href"]
        # 절대 URL일 때
        if href.startswith("https://n.news.naver.com/mnews/article"):
            full_url = href
        # 상대경로일 때
        elif href.startswith("/item/news_read.naver?article_id="):
            full_url = urljoin("https://finance.naver.com", href)
        else:
            continue
        links.append(full_url)

    # print(links)

    articles = []
    for url in links:
        try:
            art = fetch_article(url)
            articles.append(art)
        except Exception as e:
            print(f"[WARN] {url} 크롤링 중 에러: {e}")
    return articles


def fetch_article(url: str) -> dict:
    """
    finance.naver.com/item/news_read.naver?article_id=... → 
      https://n.news.naver.com/mnews/article/{office_id}/{article_id}
    로 바꿔서 크롤링
    """
    parsed = urlparse(url)

    # 1) finance.naver.com/news_read.naver → n.news.naver.com URL로 변환
    if parsed.netloc.endswith("finance.naver.com") and "news_read.naver" in parsed.path:
        qs = parse_qs(parsed.query)
        aid = qs.get("article_id", [""])[0]
        oid = qs.get("office_id", [""])[0]
        if aid and oid:
            url = f"https://n.news.naver.com/mnews/article/{oid}/{aid}"

    # 2) 실제 기사 페이지 요청
    resp = requests.get(
        url,
        headers=USER_AGENT,
        verify=certifi.where(),
        timeout=10
    )
    resp.raise_for_status()
    soup = BeautifulSoup(resp.text, "html.parser")

    # 3) 제목/본문 추출 (n.news.naver.com 공통 파서)
    title_tag = soup.select_one("h2#title_area span")
    title = title_tag.get_text(strip=True) if title_tag else ""

    article_tag = soup.select_one("article#dic_area")
    if not article_tag:
        return {"url": url, "title": "", "content": ""}

    # ① 모든 <br> 을 줄바꿈 문자열로 대체
    for br in article_tag.find_all("br"):
        br.replace_with("\n")

    # ② get_text() 에서 separator 로도 한 번 더 보정
    content = article_tag.get_text(separator="\n", strip=True)

    return {"url": url, "title": title, "content": content}

File: server/NaverNews.ts
import { load } from "cheerio";
import { httpRequest } from "./httpClient";

const DEFAULT_HEADERS = {
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64)",
  "Accept-Language": "ko-KR,ko;q=0.9,en-US;q=0.8",
};

export interface NaverArticle {
  url: string;
  title: string;
  content: string;
}

function sanitizeTickerSymbol(raw: string): string {
  return raw.replace(/\D/g, "");
}

function resolveNewsUrl(url: string): string {
  const parsed = new URL(url);

  if (parsed.hostname.endsWith("finance.naver.com") && parsed.pathname.includes("news_read.naver")) {
    const aid = parsed.searchParams.get("article_id") ?? "";
    const oid = parsed.searchParams.get("office_id") ?? "";

    if (aid && oid) {
      return `https://n.news.naver.com/mnews/article/${oid}/${aid}`;
    }
  }

  if (parsed.pathname.startsWith("/item/")) {
    return new URL(parsed.pathname + parsed.search, "https://finance.naver.com").toString();
  }

  return parsed.toString();
}

async function fetchArticle(url: string): Promise<NaverArticle> {
  const resolved = resolveNewsUrl(url);
  const response = await httpRequest(resolved, { headers: DEFAULT_HEADERS });

  if (!response.ok) {
    throw new Error(`Failed to fetch article ${resolved}: ${response.status}`);
  }

  const html = await response.text();
  const $ = load(html);

  const title = $("h2#title_area span").text().trim() || $("h2#title_area").text().trim();
  const article = $("article#dic_area");
  if (!article.length) {
    return { url: resolved, title, content: "" };
  }

  article.find("script, style, table").remove();
  article.find("br").replaceWith("\n");

  const content = article
    .text()
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean)
    .join("\n");

  return { url: resolved, title, content };
}

export async function fetchNaverNewsArticles(code: string, page = 1): Promise<NaverArticle[]> {
  const symbol = sanitizeTickerSymbol(code);
  if (!symbol) {
    throw new Error("Invalid stock code for Naver news.");
  }

  const baseUrl = `https://finance.naver.com/item/news.naver?code=${symbol}&page=${page}`;
  const initialResponse = await httpRequest(baseUrl, { headers: DEFAULT_HEADERS });

  if (!initialResponse.ok) {
    throw new Error(`Failed to fetch Naver news page: ${initialResponse.status} ${initialResponse.statusText}`);
  }

  const initialHtml = await initialResponse.text();
  const $ = load(initialHtml);

  const frameSrc = $("#news_frame").attr("src");
  if (!frameSrc) {
    return [];
  }

  const pageAwareSrc = frameSrc.includes("page=") ? frameSrc : `${frameSrc}&page=${page}`;
  const iframeUrl = new URL(pageAwareSrc, baseUrl).toString();

  const iframeHeaders = {
    ...DEFAULT_HEADERS,
    Referer: baseUrl,
  };

  const iframeResponse = await httpRequest(iframeUrl, { headers: iframeHeaders });
  if (!iframeResponse.ok) {
    throw new Error(`Failed to fetch Naver iframe: ${iframeResponse.status} ${iframeResponse.statusText}`);
  }

  const iframeHtml = await iframeResponse.text();
  const $$ = load(iframeHtml);
  $$(".relation_lst").remove();

  const links = new Set<string>();

  $$("a[href]").each((_, element) => {
    const href = $$(element).attr("href");
    if (!href) {
      return;
    }

    if (href.startsWith("https://n.news.naver.com/mnews/article")) {
      links.add(href);
      return;
    }

    if (href.startsWith("/item/news_read.naver")) {
      const absolute = new URL(href, "https://finance.naver.com").toString();
      links.add(absolute);
    }
  });

  const articles: NaverArticle[] = [];

  for (const url of links) {
    try {
      const article = await fetchArticle(url);
      articles.push(article);
    } catch (error) {
      console.warn(`[WARN] Failed to crawl article ${url}: ${(error as Error).message}`);
    }
  }

  return articles;
}
